New violation: Prefer TypeError for unexpected types

swellander opened the pull request:
This currently only covers the one example case outlined by @sbrugman in [issue #32](https://github.com/guilatrova/tryceratops/issues/32). I want to make sure that I'm on the right track. Thanks!

guilatrova commented on src/tests/samples/violations/conditional_prefer_type_error.py:
```diff
@@ -0,0 +1,176 @@
+        raise Exception("...")  # should be typeerror
+
+
+def incorrect_multiple_operators(some_arg):
+    if (False and (3 == 3)) or (not isinstance(some_arg, int) and True):
+        pass
+    else:
+        raise Exception("...")  # should be typeerror
```
I guess that will be tough to ensure we understood the developer's intention. (We might get it sometimes, and miss other times).

I'd prefer to remove it.
Thoughts? @sbrugman

sbrugman replied:
Boolean functions including isinstance are endless and not guaranteed to be execlusively type checking. I'd opt for restricting ourselves to the clear-cut cases.

guilatrova replied:
Yes, **strict cleat-cut cases** like:

- `if isinstance()`
- `if not isinstance()`
- `if isinstance() and isinstance()`
- `if isinstance() and not isinstance()`
- `if isinstance() and not isinstance() and isinstance()`
- ...

(and variations you did e.g. `callable`, `issubclass`, etc...)

WDYT @swellander ?

swellander replied:
That makes sense to me 👍

guilatrova commented on src/tryceratops/analyzers/conditional.py:
```diff
@@ -0,0 +1,62 @@
+
+from .base import BaseAnalyzer
+
+STANDARD_NON_TYPE_ERROR_IDS = (
+    "ArithmeticError",
+    "AssertionError",
+    "AttributeError",
+    "BufferError",
+    "EOFError",
+    "Exception",
+    "ImportError",
+    "LookupError",
+    "MemoryError",
+    "NameError",
+    "ReferenceError",
+    "RuntimeError",
+    "SyntaxError",
+    "SystemError",
+    "ValueError",
+)
```
Nice, that's interesting, let's try!

I'll run this version on some personal projects and see the output. I'm excited!

guilatrova commented on src/tests/samples/violations/conditional_prefer_type_error.py:
```diff
@@ -0,0 +1,176 @@
+        raise Exception("...")  # should be typeerror
+
+
+def incorrect_elif(some_arg):
```
Great!

sbrugman replied:
This is not as obvious to me. I'd suggest to restricting where _all_ conditions consist of one of the type checking functions

guilatrova replied:
After reviewing it again, I agree with @sbrugman 

It's hard to tell the exception being raised there was caused by the type checking.

swellander replied:
Agreed!

guilatrova commented on src/tryceratops/analyzers/conditional.py:
```diff
@@ -0,0 +1,62 @@
+            for stm in ast.iter_child_nodes(node):
+                if isinstance(stm, ast.Raise):
+                    self._check_is_raise_other_than_typeerror(stm)
+        else:
```
I think this `else` block is not required at all.
I'll run some tests locally later, and give feedback.

guilatrova replied:
Indeed, instead of an `else`, invoke `generic_visit` again:

```py
    def visit_If(self, node: ast.If) -> None:
        if self._is_checking_type(node):
            for stm in ast.iter_child_nodes(node):
                if isinstance(stm, ast.Raise):
                    self._check_is_raise_other_than_typeerror(stm)

        self.generic_visit(node) # <- Here
```

swellander replied:
Good catch. I ended up moving the recursive call into `_is_checking_type()` instead so that we could make sure all conditional branches involve some kind of type-check, as per [@sbrugman's earlier comment](https://github.com/guilatrova/tryceratops/pull/34#discussion_r740425891), before marking a violation.

guilatrova commented:
We're in great shape here! Feel free to convert from a draft to a real PR.

Great work, I'm impressed by:

- How clear your code is
- Great commit history (they'll be pretty on the changelog) 
- Following the conventions and organization followed by the project

Lastly, great work on rebasing your git history without I even asking. You're clearly a great developer! 👏 
I'm glad that I had your help to push forward this nice feature to keep python try/except pretty ✨

sbrugman commented on src/tests/samples/violations/conditional_prefer_type_error.py:
```diff
@@ -0,0 +1,176 @@
+
+
+def incorrect_ArithmeticError(some_arg):
+    if isinstance(some_arg, int):
```
Perhaps parametrize all these tests using the exception types for simplicity

guilatrova reviewed (changes requested):
I run it on some private code I have here, and I'm surprised it catch some violations that I implemented haha (which is good!).

There was one case that failed though (fake alert), let me share so we can start discussing:

### Case 1 (failing)

```py
if val is not None and (
    not isinstance(val, int) or val < 0
):
    raise ValueError(...)  # <- Got alert
```

I think for this case we can't mark errors because the condition is complex and contains validation unrelated to type.

Which is in sync with out discussion: https://github.com/guilatrova/tryceratops/pull/34#discussion_r740424720

Also, let's have a unit test to validate we ignore this scenario.

Thoughts? @swellander 

### Case 2 (Worked, but we should have a test)

Example 1: (We should support)

```py
if isinstance(val1, type1) and isinstance(val2, type2):
    ...
else:
    raise ValueError("...")
```

Example 2: (It would be nice to support, I'm ok if we skip it for now)

```py
if not (isinstance(sources, list) and
                all(isinstance(v, str) for v in sources)):
            raise AssertionError("'sources' must be a list of strings")
```

If ALL conditions for an if are checking the type, then we mark as violation!

---

## Remove a useless `else`

As mentioned here: https://github.com/guilatrova/tryceratops/pull/34#discussion_r748700487

---

Great work so far! Let me know if you have the bandwidth to finish the implementation as suggested above @swellander

guilatrova commented:
@swellander How's your time? Do you think you will be able to make the final adjustments?

swellander commented:
@guilatrova yes, sorry again for the delay! I will have time to submit work tomorrow night if that works for you. If not, I don't want to hold up the project

swellander commented on src/tests/samples/violations/conditional_prefer_type_error.py:
```diff
@@ -0,0 +1,258 @@
+        raise ValueError(...)  # fine if this is not a TypeError
+
+
+def correct_multi_conditional(some_arg):
+    if some_arg == 3:
+        pass
+    elif isinstance(some_arg, int):
+        pass
+    else:
+        raise Exception("...")  # fine if this is not a TypeError
```
This case should **not** trigger a violation because not all conditions check for a type

swellander commented on src/tests/samples/violations/conditional_prefer_type_error.py:
```diff
@@ -0,0 +1,258 @@
+        raise MyCustomTypeValidation("...")  # that's correct, because it's not vanilla
+
+
+def correct_complex_conditional(val):
+    if val is not None and (not isinstance(val, int) or val < 0):
+        raise ValueError(...)  # fine if this is not a TypeError
```
@guilatrova this is testing Case 2, example 1 from your most recent PR review

swellander commented on src/tests/samples/violations/conditional_prefer_type_error.py:
```diff
@@ -0,0 +1,258 @@
+def incorrect_multiple_operators_callable(arg1, arg2, arg3):
+    if not callable(arg1) and callable(arg2) or callable(arg3):
+        pass
+    else:
+        raise Exception("...")  # should be typeerror
+
+
+def incorrect_not_operator_issubclass(some_arg):
+    if not issubclass(some_arg):
+        pass
+    else:
+        raise Exception("...")  # should be typeerror
+
+
+def incorrect_and_operator_issubclass(arg1, arg2):
+    if issubclass(some_arg) and issubclass(arg2):
+        pass
+    else:
+        raise Exception("...")  # should be typeerror
+
+
+def incorrect_or_operator_issubclass(arg1, arg2):
+    if issubclass(some_arg) or issubclass(arg2):
+        pass
+    else:
+        raise Exception("...")  # should be typeerror
+
+
+def incorrect_multiple_operators_issubclass(arg1, arg2, arg3):
+    if not issubclass(arg1) and issubclass(arg2) or issubclass(arg3):
+        pass
+    else:
+        raise Exception("...")  # should be typeerror
+
+
+def incorrect_multi_conditional(arg1, arg2):
+    if isinstance(arg1, int):
+        pass
+    elif isinstance(arg2, int):
+        raise Exception("...")  # should be typeerror
```
Added tests for the [clear-cut cases we discussed](https://github.com/guilatrova/tryceratops/pull/34#discussion_r748702026)

swellander commented:
@guilatrova @sbrugman I think I've addressed all your comments **except** for [the parametrization of tests](https://github.com/guilatrova/tryceratops/pull/34#discussion_r740427039) and the Case 2, Example 2 from your review, @guilatrova. I hope to get to those two changes tomorrow night. Thanks!

swellander commented on src/tests/analyzers_conditional_test.py:
```diff
@@ -0,0 +1,58 @@
+        assert_violation, codes.PREFER_TYPE_ERROR[0], codes.PREFER_TYPE_ERROR[1]
+    )
+
+    for non_type_exception in STANDARD_NON_TYPE_ERROR_IDS:
+        with tempfile.NamedTemporaryFile(mode="r+") as tmp:
+            file_content = (
+                "if isinstance(some_arg, int):\n"
+                "   pass\n"
+                "else:\n"
+                f'   raise {non_type_exception}("...")'
+            )
+            tmp.write(file_content)
+            tmp.seek(0)
+            content = tmp.read()
+
+            tmp_tree = ast.parse(content)
+            tmp_violations = analyzer.check(tmp_tree, "filename")
+
+            assert len(tmp_violations) == 1
+            assert_type_error(4, 9, tmp_violations[0])
```
@sbrugman is this what you had in mind when you mentioned [parametrizing these tests](https://github.com/guilatrova/tryceratops/pull/34#discussion_r740427039)?

guilatrova replied:
I think he meant to try [`pytest.mark.parametrize`](https://docs.pytest.org/en/6.2.x/parametrize.html). But I'm not confident it would make the code any easier.

swellander replied:
Ah I see, thanks

guilatrova commented:
> @guilatrova @sbrugman I think I've addressed all your comments except for the parametrization of tests and the Case 2, Example 2 from your review, @guilatrova. I hope to get to those two changes tomorrow night. Thanks!

@swellander 
I'd recommend the parametrization of tests and case 2 to be done later in future PRs, to keep this one shorter.
WDYT? Also, I'm excited to merge this feature, so let me know your thoughts.

Once the commits are "done", then I'll run on some private repos to sanity check.

guilatrova commented:
### Versioninig question

@sbrugman @swellander 
Another important thought, should we consider this release a "breaking change"?

I'm thinking it should. Mostly because anyone that was using the `0.6.x` version and tryceratops was passing, maybe now will start having problems. I'd feel more comfortable making it a version `1.x` now.

### Suggestion

If you agree, @swellander do you mind making your commit `feat: add prefer TypeError analyzer` a breaking change by adding the `!`? (e.g. `feat!: add prefer TypeError analyzer`).

swellander commented:
> I'd recommend the parametrization of tests and case 2 to be done later in future PRs, to keep this one shorter.
WDYT? Also, I'm excited to merge this feature, so let me know your thoughts.
@guilatrova saving those changes for a new PR sounds good 👍 

And treating this feature as a breaking change makes sense to me! I'll use the breaking change syntax when I squash the commits together here shortly.

guilatrova commented:
I just tested here against some real code and it looks working.

guilatrova commented:
I don't have enough words to express my gratitude.

Thank you @sbrugman and @swellander for helping the community and for contributing to making everyone's Python code better ❤️ ✨

swellander commented:
Thank you @sbrugman for the wise input and thank you @guilatrova for the helpful feedback and your positive, welcoming energy 🌞 !
